src/Planner/PlannerContext.cpp

```diff
@@ -120,9 +121,24 @@ const ColumnIdentifier * PlannerContext::getColumnNodeIdentifierOrNull(const Que
     return table_expression_data->getColumnIdentifierOrNull(column_name);
 }
 
-PlannerContext::SetKey PlannerContext::createSetKey(const QueryTreeNodePtr & set_source_node)
+PlannerContext::SetKey PlannerContext::createSetKey(const DataTypePtr & left_operand_type, const QueryTreeNodePtr & set_source_node)
 {
     auto set_source_hash = set_source_node->getTreeHash();
+
+    if (set_source_node->as<ConstantNode>())
+    {
+        /* We need to hash the type of the left operand because we can build different sets for different types.
+         * (It's done for performance reasons. It's cheaper to convert a small set of values from literal to the type of the left operand.)
+         *
+         * For example in expression `(a :: Decimal(9, 1) IN (1.0, 2.5)) AND (b :: Decimal(9, 0) IN (1, 2.5))`
+         * we need to build two different sets:
+         *   - `{1, 2.5} :: Set(Decimal(9, 1))` for a
+         *   - `{1} :: Set(Decimal(9, 0))` for b (2.5 omitted because bercause it's not representable as Decimal(9, 0)).
+         */
+        return "__set_" + left_operand_type->getName() + '_' + toString(set_source_hash.first) + '_' + toString(set_source_hash.second);
```

Comment:
I guess the correct way is to use a common type, not the left type. What would happen with `1.0 IN ((a :: Decimal(9, 1)))`?

Reply:
For constant in right side of `IN` we build set of type of left operand
(`first_argument_constant_type` -> `set_element_types` -> `block_types`)

https://github.com/ClickHouse/ClickHouse/blob/5ecdb3502908b5148c081cebc7ababc2a84ec173/src/Analyzer/Passes/QueryAnalysisPass.cpp#L5120-L5121

https://github.com/ClickHouse/ClickHouse/blob/5ecdb3502908b5148c081cebc7ababc2a84ec173/src/Analyzer/SetUtils.cpp#L159-L160

https://github.com/ClickHouse/ClickHouse/blob/5ecdb3502908b5148c081cebc7ababc2a84ec173/src/Analyzer/SetUtils.cpp#L114-L115